src/Disks/DiskCacheWrapper.cpp

```diff
@@ -0,0 +1,225 @@
+#include "DiskCacheWrapper.h"
+#include <IO/copyData.h>
+#include <Common/quoteString.h>
+#include <common/logger_useful.h>
+
+
+namespace DB
+{
+DiskCacheWrapper::DiskCacheWrapper(
+    std::shared_ptr<IDisk> delegate_, std::shared_ptr<DiskLocal> cache_disk_, std::function<bool(const String &)> cache_file_predicate_)
+    : DiskDecorator(delegate_), cache_disk(cache_disk_), cache_file_predicate(cache_file_predicate_)
+{
+}
+
+std::shared_ptr<FileDownloadMetadata> DiskCacheWrapper::acquireDownloadMetadata(const String & path) const
+{
+    std::unique_lock<std::mutex> lock{mutex};
+
+    auto it = file_downloads.find(path);
+    if (it != file_downloads.end() && !it->second.expired())
+        return it->second.lock();
```

Comment:
`lock()` can return a nullptr when the object is deleted, looks like you don't check for that

Reply:
I checked it by `expired()` call